File: backend/src/main/java/com/bobcarrental/service/ClientService.java
package com.bobcarrental.service;

import com.bobcarrental.dto.request.ClientRequest;
import com.bobcarrental.dto.response.ClientResponse;
import com.bobcarrental.dto.response.ClientSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Client service interface.
 * Handles all business logic related to client management.
 */
public interface ClientService {
    
    /**
     * Create a new client
     */
    ClientResponse createClient(ClientRequest request);
    
    /**
     * Update an existing client
     */
    ClientResponse updateClient(Long id, ClientRequest request);
    
    /**
     * Get client by database ID
     */
    ClientResponse getClientById(Long id);
    
    /**
     * Get client by legacy client ID
     */
    ClientResponse getClientByClientId(String clientId);
    
    /**
     * Delete a client
     */
    void deleteClient(Long id);
    
    /**
     * Get all clients with pagination
     */
    Page<ClientSummaryResponse> getAllClients(Pageable pageable);
    
    /**
     * Search clients by name, ID, city, or phone
     */
    List<ClientSummaryResponse> searchClients(String search);
    
    /**
     * Get clients by city
     */
    List<ClientSummaryResponse> getClientsByCity(String city);
    
    /**
     * Validate client ID uniqueness (PresenceChk from legacy)
     */
    
    /**
     * Search clients by name
     */
    List<ClientSummaryResponse> searchClientsByName(String name);
    
    /**
     * Check if client name exists
     */
    boolean existsByName(String name);
    
    /**
     * Check if phone number exists
     */
    boolean existsByPhone(String phone);
    void validateClientId(String clientId, Long excludeId);
    
    /**
     * Count active clients
     */
    long countActiveClients();
}

// Made with Bob

File: frontend/src/app/core/services/client.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Client, ClientRequest, ClientResponse, ApiResponse } from '../../models';

interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  first: boolean;
  last: boolean;
  hasPrevious: boolean;
  hasNext: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class ClientService {
  private readonly API_URL = `${environment.apiUrl}/clients`;

  constructor(private http: HttpClient) {}

  /**
   * Get all clients
   */
  getAllClients(): Observable<ClientResponse[]> {
    return this.http.get<ApiResponse<PageResponse<ClientResponse>>>(this.API_URL).pipe(
      map(response => response.data.content)
    );
  }

  /**
   * Get client by ID
   */
  getClientById(id: number): Observable<ClientResponse> {
    return this.http.get<ApiResponse<ClientResponse>>(`${this.API_URL}/${id}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Search clients by name
   */
  searchClients(name: string): Observable<ClientResponse[]> {
    const params = new HttpParams().set('name', name);
    return this.http.get<ApiResponse<ClientResponse[]>>(`${this.API_URL}/search`, { params }).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get active clients only
   */
  getActiveClients(): Observable<ClientResponse[]> {
    return this.http.get<ApiResponse<ClientResponse[]>>(`${this.API_URL}/active`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Create new client
   */
  createClient(request: ClientRequest): Observable<ClientResponse> {
    return this.http.post<ApiResponse<ClientResponse>>(this.API_URL, request).pipe(
      map(response => response.data)
    );
  }

  /**
   * Update existing client
   */
  updateClient(id: number, request: ClientRequest): Observable<ClientResponse> {
    return this.http.put<ApiResponse<ClientResponse>>(`${this.API_URL}/${id}`, request).pipe(
      map(response => response.data)
    );
  }

  /**
   * Delete client
   */
  deleteClient(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.API_URL}/${id}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get client statistics
   */
  getClientStats(clientId: number): Observable<any> {
    return this.http.get<ApiResponse<any>>(`${this.API_URL}/${clientId}/stats`).pipe(
      map(response => response.data)
    );
  }
}

// Made with Bob
